source/common/stats/stats_impl.h

```diff
@@ -473,5 +477,131 @@ class IsolatedStoreImpl : public Store {
   IsolatedStatsCache<Histogram, HistogramImpl> histograms_;
 };
 
+/**
+ * Structure to hold statistical data of histogram.
+ */
+struct HistogramStatistics {
+public:
+  HistogramStatistics(const std::string name, histogram_t* histogram_ptr) : name_(name) {
+    double arr[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
+    double out[9];
+    double* arrptr = arr;
+    double* outptr = out;
+    hist_approx_quantile(histogram_ptr, arrptr, 9, outptr);
+    p0_ = out[0];
+    p25_ = out[1];
+    p50_ = out[2];
+    p75_ = out[3];
+    p90_ = out[4];
+    p95_ = out[5];
+    p99_ = out[6];
+    p999_ = out[7];
+    p100_ = out[8];
+  }
+
+  std::string name() const { return name_; }
+
+  std::string summary() const {
+    return fmt::format("P0: {} , P25: {}, P50: {}, P75: {}, P90: {}, P95: {}, P99: {}, P100: {}",
+                       p0_, p25_, p50_, p75_, p90_, p95_, p99_, p999_, p100_);
+  }
+
+private:
+  const std::string name_;
+  double p0_;
+  double p25_;
+  double p50_;
+  double p75_;
+  double p90_;
+  double p95_;
+  double p99_;
+  double p999_;
+  double p100_;
+};
+
+/**
+ * BaseSink implementation for all Sinks. This mainly  .
+ */
+class BaseSink : public Sink {
+public:
+  void onHistogramComplete(const Histogram& hist, uint64_t value) override {
+    ThreadLocalHistograms tls_histograms = hist_tls_->getTyped<ThreadLocalHistograms>();
+    auto tls_hist_iterator = tls_histograms.histograms_.find(hist.name());
+    if (tls_hist_iterator != tls_histograms.histograms_.end()) {
+      hist_insert(tls_hist_iterator->second, value, 1);
+    } else {
+      histogram_t* histogram_ptr = hist_alloc();
```

Comment:
Please use https://github.com/envoyproxy/envoy/blob/master/source/common/common/c_smart_ptr.h for this.

Also, I think all of this function needs to be locked. You can probably use a lock per thread, but this needs to be synchronized with merging (see more below on threading).

Reply:
sure. I have not done much about locking and synchronization in the first pass. My idea is to build the entire thing first and comeback to locking/sync for onHistogramComplete and onFlushHistograms at the end so that we can focus completely on these two functions and optimize it.